sdk/include/opentelemetry/sdk/common/http_client.h

```diff
@@ -0,0 +1,129 @@
+#pragma once
+
+#include <chrono>
+#include <string>
+#include <vector>
+#include "opentelemetry/nostd/function_ref.h"
+#include "opentelemetry/version.h"
+
+// Usage Example
+//
+// struct ReponseHandler: public HttpResponseHandler {
+//      void OnHttpResponse(HttpResponse& res) noexcept override
+//      {
+//           if (res.IsSuccess())
+//              .. process response
+//      }
+//  };
+//
+//   HttpSessionManager sessionManager; // implementer can provide singleton implementation for it
+//   auto session = sessionManager.createHttpSession("localhost", 8000);
+//   auto request = session->CreateRequest();
+//   request->AddHeader(..);
+//   ResponseHandler res_handler;
+//   session->SendRequest(res_handler);
+//   session->FinishSession() // optionally in the end.
+//   ...shutdown
+//    sessionManager.FinishAllSessions()
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace common
+{
+namespace http
+{
+
+enum class HttpMethod
+{
+  HTTP_GET,
+  HTTP_POST,
+  HTTP_PUT,
+  HTTP_OPTIONS,
+  HTTP_HEAD,
+  HTTP_PATCH,
+  HTTP_DELETE
+};
+
+enum class HttpSessionState
+{
+  CREATED,  // session object is created
+  ONGOING,  // http(s) request is ongoing
+  FINISHED  // https(s) request is finished
+};
+
+using HttpHeader     = std::pair<std::string, std::string>;
+using HttpBody       = std::vector<uint8_t>;
+using HttpStatusCode = uint16_t;
+
+class HttpRequest
+{
+
+public:
+  virtual void setMethod(HttpMethod method) noexcept = 0;
+
+  virtual void setURI(const std::string &uri) noexcept = 0;
+
+  virtual void SetBody(HttpBody &body) noexcept = 0;
+
+  virtual void AddHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void ReplaceHeader(const HttpHeader &header) noexcept = 0;
+
+  virtual void SetTimeoutMs(std::chrono::milliseconds timeout_ms) noexcept = 0;
+};
+
+class HttpResponse
+{
+public:
+  virtual const HttpBody &GetBody() const noexcept = 0;
+
+  virtual bool GetNextHeader(nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual bool GetNextHeader(const std::string &key,
+                             nostd::function_ref<bool(HttpHeader)> callable) const noexcept = 0;
+
+  virtual HttpStatusCode GetStatusCode() const noexcept = 0;
+
+  virtual bool IsSuccess() const noexcept = 0;
+
+  virtual const std::string &GetErrorMessage() const noexcept = 0;
+};
+
+class HttpResponseHandler
+{
+
+public:
+  virtual void OnHttpResponse(HttpResponse &) noexcept = 0;
+};
+
+class HttpSession
+{
+
+public:
+  virtual std::shared_ptr<HttpRequest> CreateRequest() noexcept = 0;
+
+  virtual void SendRequest(HttpResponseHandler &) noexcept = 0;
+
+  virtual HttpSessionState GetSessionState() const noexcept = 0;
```

Comment:
Can separate asynchronous requests be in progress for a single session? If yes, whould `GetSessionState` return an aggregate status for those cases?

Reply:
Current API design doesn't allows to have multiple async http requests in single session. While this is definitely a good use-case to have, as of now this is not supported in API.

Reply:
👍 

It would be great to document that. Also: in case one sends a request while another request is in progress, will there be an error message? Will an `HttpResponse` with `IsSuccess() == false` be passed to the handler?

Reply:
There are two scenarios
 - Calling `createRequest()` second time for give session would return `nullptr`.
- Calling `sendRequest()` second time for given session would invoke `handler::onError()`.

`GetSessionState`, `CancelSession`, `FinishSession` are actionable on first request.